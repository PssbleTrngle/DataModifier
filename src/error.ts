import { Logger } from './logger.js'
import { ZodError } from 'zod'
import { NormalizedId } from './common/id.js'

export class IllegalShapeError extends Error {
   constructor(message: string, readonly input?: unknown) {
      super(message)
   }
}

export class UnknownRegistryEntry extends Error {
   constructor(message: string, readonly registry: string, readonly id: NormalizedId) {
      super(message)
   }
}

export function tryCatching<T>(logger: Logger | undefined, run: () => T): T | null {
   try {
      return run()
   } catch (error) {
      if (error instanceof IllegalShapeError) {
         if (error.input) logger?.warn(error.message, error.input)
         else logger?.warn(error.message)
         return null
      }

      if (error instanceof ZodError) {
         const message = error.errors
            .map(it => {
               if (it.path) return `${it.path.join('.')}: ${it.message}`
               else return it.message
            })
            .join(', ')
         logger?.warn(`unknown shape: ${message}`)
         return null
      }

      throw error
   }
}
