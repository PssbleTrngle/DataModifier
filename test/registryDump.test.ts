import { createIngredient, createResult, PackLoader } from '../src/index.js'
import createTestLogger from './mock/TestLogger.js'
import { createResolver } from '@pssbletrngle/pack-resolver'
import { tryCatching } from '../src/error.js'

const logger = createTestLogger()
const loader = new PackLoader(logger)

beforeEach(async () => {
   const resolver = createResolver({ from: 'test/resources/dump', silent: true })
   await loader.loadRegistryDump(resolver)
})

afterEach(() => {
   logger.reset()
   loader.clear()
})

describe('registry dump tests', () => {
   it('correctly loads imports registries', async () => {
      expect(logger.warn).not.toHaveBeenCalled()
      expect(loader.registries.keys('fluid')).toMatchSnapshot('dumped fluid registry entries')
   })

   it('warns about unknown registries', async () => {
      loader.registries.keys('example')

      expect(logger.warn).toHaveBeenCalledWith(
         "tried to access registry 'minecraft:example', which has not been loaded"
      )
   })

   it('validates correct ingredients', async () => {
      createIngredient('minecraft:stone', loader.registries)
      createResult({ item: 'minecraft:stick' }, loader.registries)
      createIngredient({ block: 'minecraft:obsidian' }, loader.registries)
      createResult({ fluid: 'minecraft:water' }, loader.registries)

      expect(logger.error).not.toHaveBeenCalled()
      expect(logger.warn).not.toHaveBeenCalled()
   })

   it('validates incorrect ingredients', async () => {
      expect(() => tryCatching(logger, () => createIngredient('example:unknown', loader.registries))).toThrow(
         "unknown minecraft:item 'example:unknown'"
      )

      expect(() => tryCatching(logger, () => createResult({ item: 'minecraft:kitkat' }, loader.registries))).toThrow(
         "unknown minecraft:item 'minecraft:kitkat'"
      )

      expect(() => tryCatching(logger, () => createIngredient({ block: 'something' }, loader.registries))).toThrow(
         "unknown minecraft:block 'minecraft:something'"
      )

      expect(() => tryCatching(logger, () => createResult({ fluid: 'whatever' }, loader.registries))).toThrow(
         "unknown minecraft:fluid 'minecraft:whatever'"
      )

      expect(() =>
         tryCatching(logger, () => createIngredient(['minecraft:stone', { fluid: 'no-idea' }], loader.registries))
      ).toThrow("unknown minecraft:fluid 'minecraft:no-idea'")
   })
})
