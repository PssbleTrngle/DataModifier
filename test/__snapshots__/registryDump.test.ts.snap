// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registry dump tests correctly loads imports registries: dumped fluid registry entries 1`] = `
Set {
  "minecraft:lava",
  "minecraft:empty",
  "minecraft:water",
  "minecraft:flowing_lava",
  "minecraft:flowing_water",
}
`;
